File: js/plugins/vertexai/tests/plugin_test.ts
/**
 * Copyright 2024 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as assert from 'assert';
import { genkit } from 'genkit';
import type { ModelInfo } from 'genkit/model';
import { describe, it } from 'node:test';
import { __setFakeDerivedParams } from '../src/common/index.js';
import { GENERIC_GEMINI_MODEL, gemini } from '../src/gemini.js';
import vertexAI, { gemini15Flash, gemini15Pro } from '../src/index.js';

describe('plugin', () => {
  __setFakeDerivedParams({
    projectId: 'test',
    location: 'us-central1',
  });

  it('should init the plugin without requiring the api key', async () => {
    const ai = genkit({
      plugins: [vertexAI()],
    });

    assert.ok(ai);
  });

  it('should pre-register a few flagship models', async () => {
    const ai = genkit({
      plugins: [vertexAI()],
    });

    assert.ok(await ai.registry.lookupAction(`/model/${gemini15Flash.name}`));
    assert.ok(await ai.registry.lookupAction(`/model/${gemini15Pro.name}`));
  });

  it('allow referencing models using `gemini` helper', async () => {
    const ai = genkit({
      plugins: [vertexAI()],
    });

    const pro = await ai.registry.lookupAction(
      `/model/${gemini('gemini-1.5-pro').name}`
    );
    assert.ok(pro);
    assert.strictEqual(pro.__action.name, 'vertexai/gemini-1.5-pro');
    const flash = await ai.registry.lookupAction(
      `/model/${gemini('gemini-1.5-flash').name}`
    );
    assert.ok(flash);
    assert.strictEqual(flash.__action.name, 'vertexai/gemini-1.5-flash');
  });

  it('references dynamic models', async () => {
    const ai = genkit({
      plugins: [vertexAI({ location: 'us-central1' })],
    });
    const giraffeRef = gemini('gemini-4.5-giraffe');
    assert.strictEqual(giraffeRef.name, 'vertexai/gemini-4.5-giraffe');
    const giraffe = await ai.registry.lookupAction(`/model/${giraffeRef.name}`);
    assert.ok(giraffe);
    assert.strictEqual(giraffe.__action.name, 'vertexai/gemini-4.5-giraffe');
    assertEqualModelInfo(
      giraffe.__action.metadata?.model,
      'Vertex AI - gemini-4.5-giraffe',
      GENERIC_GEMINI_MODEL.info! // <---- generic model fallback
    );
  });

  it('references pre-registered models', async () => {
    const flash002Ref = gemini('gemini-1.5-flash-002');
    const ai = genkit({
      plugins: [
        vertexAI({
          location: 'us-central1',
          models: ['gemini-1.5-pro-002', flash002Ref, 'gemini-4.0-banana'],
        }),
      ],
    });

    const pro002Ref = gemini('gemini-1.5-pro-002');
    assert.strictEqual(pro002Ref.name, 'vertexai/gemini-1.5-pro-002');
    assertEqualModelInfo(
      pro002Ref.info!,
      'Vertex AI - gemini-1.5-pro-002',
      gemini15Pro.info!
    );
    const pro002 = await ai.registry.lookupAction(`/model/${pro002Ref.name}`);
    assert.ok(pro002);
    assert.strictEqual(pro002.__action.name, 'vertexai/gemini-1.5-pro-002');
    assertEqualModelInfo(
      pro002.__action.metadata?.model,
      'Vertex AI - gemini-1.5-pro-002',
      gemini15Pro.info!
    );

    assert.strictEqual(flash002Ref.name, 'vertexai/gemini-1.5-flash-002');
    assertEqualModelInfo(
      flash002Ref.info!,
      'Vertex AI - gemini-1.5-flash-002',
      gemini15Flash.info!
    );
    const flash002 = await ai.registry.lookupAction(
      `/model/${flash002Ref.name}`
    );
    assert.ok(flash002);
    assert.strictEqual(flash002.__action.name, 'vertexai/gemini-1.5-flash-002');
    assertEqualModelInfo(
      flash002.__action.metadata?.model,
      'Vertex AI - gemini-1.5-flash-002',
      gemini15Flash.info!
    );

    const bananaRef = gemini('gemini-4.0-banana');
    assert.strictEqual(bananaRef.name, 'vertexai/gemini-4.0-banana');
    assertEqualModelInfo(
      bananaRef.info!,
      'Vertex AI - gemini-4.0-banana',
      GENERIC_GEMINI_MODEL.info! // <---- generic model fallback
    );
    const banana = await ai.registry.lookupAction(`/model/${bananaRef.name}`);
    assert.ok(banana);
    assert.strictEqual(banana.__action.name, 'vertexai/gemini-4.0-banana');
    assertEqualModelInfo(
      banana.__action.metadata?.model,
      'Vertex AI - gemini-4.0-banana',
      GENERIC_GEMINI_MODEL.info! // <---- generic model fallback
    );
  });
});

function assertEqualModelInfo(
  modelAction: ModelInfo,
  expectedLabel: string,
  expectedInfo: ModelInfo
) {
  assert.strictEqual(modelAction.label, expectedLabel);
  assert.deepStrictEqual(modelAction.supports, expectedInfo.supports);
  assert.deepStrictEqual(modelAction.versions, expectedInfo.versions);
}

File: py/plugins/anthropic/tests/plugin_test.py
# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# SPDX-License-Identifier: Apache-2.0

"""Tests for Anthropic plugin."""

from unittest.mock import patch

import pytest

from genkit.core.registry import ActionKind
from genkit.plugins.anthropic import Anthropic, anthropic_name
from genkit.plugins.anthropic.model_info import (
    SUPPORTED_ANTHROPIC_MODELS as SUPPORTED_MODELS,
    get_model_info,
)
from genkit.types import (
    GenerateRequest,
    GenerationCommonConfig,
    Message,
    Part,
    Role,
    TextPart,
    ToolDefinition,
)


def test_anthropic_name() -> None:
    """Test anthropic_name helper function."""
    assert anthropic_name('claude-sonnet-4') == 'anthropic/claude-sonnet-4'


def test_init_with_api_key() -> None:
    """Test plugin initialization with API key."""
    plugin = Anthropic(api_key='test-key')
    assert plugin._anthropic_client.api_key == 'test-key'
    assert plugin.models == list(SUPPORTED_MODELS.keys())


def test_init_without_api_key_raises() -> None:
    """Test plugin initialization without API key uses default behavior."""
    with patch.dict('os.environ', {}, clear=True):
        # AsyncAnthropic allows initialization without API key
        # Error only occurs when making actual API calls
        plugin = Anthropic()
        assert plugin._anthropic_client is not None


def test_init_with_env_var() -> None:
    """Test plugin initialization with environment variable."""
    with patch.dict('os.environ', {'ANTHROPIC_API_KEY': 'env-key'}):
        plugin = Anthropic()
        assert plugin._anthropic_client.api_key == 'env-key'


def test_custom_models() -> None:
    """Test plugin initialization with custom models."""
    plugin = Anthropic(api_key='test-key', models=['claude-sonnet-4'])
    assert plugin.models == ['claude-sonnet-4']


@pytest.mark.asyncio
async def test_plugin_init() -> None:
    """Test plugin init method."""
    plugin = Anthropic(api_key='test-key', models=['claude-sonnet-4'])

    # init() should return an empty list (using lazy loading)
    result = await plugin.init()
    assert result == []


@pytest.mark.asyncio
async def test_resolve_action_model() -> None:
    """Test resolve method for model."""
    plugin = Anthropic(api_key='test-key')

    # Test resolving with unprefixed name
    action = await plugin.resolve(ActionKind.MODEL, 'anthropic/claude-sonnet-4')

    assert action is not None
    assert action.name == 'anthropic/claude-sonnet-4'
    assert action.kind == ActionKind.MODEL


def test_supported_models() -> None:
    """Test that all supported models have proper metadata."""
    assert len(SUPPORTED_MODELS) == 7
    for _name, info in SUPPORTED_MODELS.items():
        assert info.label is not None
        assert info.label.startswith('Anthropic - ')
        assert info.versions is not None
        assert len(info.versions) > 0
        assert info.supports is not None
        assert info.supports.multiturn is True
        assert info.supports.tools is True
        assert info.supports.media is True
        assert info.supports.system_role is True


def test_get_model_info_known() -> None:
    """Test get_model_info returns correct info for known model."""
    info = get_model_info('claude-sonnet-4')
    assert info.label == 'Anthropic - Claude Sonnet 4'
    assert info.supports is not None
    assert info.supports.multiturn is True
    assert info.supports.tools is True


def test_get_model_info_unknown() -> None:
    """Test get_model_info returns default info for unknown model."""
    info = get_model_info('unknown-model')
    assert info.label == 'Anthropic - unknown-model'
    assert info.supports is not None
    assert info.supports.multiturn is True
    assert info.supports.tools is True


def _create_sample_request() -> GenerateRequest:
    """Create a sample generation request for testing."""
    return GenerateRequest(
        messages=[
            Message(
                role=Role.USER,
                content=[Part(root=TextPart(text='Hello, how are you?'))],
            )
        ],
        config=GenerationCommonConfig(),
        tools=[
            ToolDefinition(
                name='get_weather',
                description='Get weather for a location',
                input_schema={
                    'type': 'object',
                    'properties': {'location': {'type': 'string', 'description': 'Location name'}},
                    'required': ['location'],
                },
            )
        ],
    )
